fix a few errant `Krate` edges

nikomatsakis opened the pull request:
Exploring the effect of small changes on `syntex` reuse, I discovered the following sources of unnecessary edges from `Krate`

r? @michaelwoerister

nagisa commented on src/librustc/dep_graph/graph.rs:
```diff
@@ -38,6 +38,10 @@ struct DepGraphData {
+
+    /// Tracks nodes that are forbidden to read/write; see
+    /// `DepGraph::forbid`. Used for debugging only.
+    forbidden: RefCell<Vec<DepNode<DefId>>>,
```
Should this field also be #[cfg(debug_assertions)]-ed?

michaelwoerister commented on src/test/incremental/krate-inherent.rs:
```diff
@@ -0,0 +1,34 @@
+
+    fn method() {
+        let x: Foo = Foo;
+        x.foo(); // inherent methods used to add an edge from Krate
```
Wow, this is really `:)` How did anything ever get re-used before this fix.

michaelwoerister commented on src/librustc/hir/map/mod.rs:
```diff
@@ -928,6 +928,8 @@ pub fn map_decoded_item<'ast, F: FoldOps>(map: &Map<'ast>,
                                           ii: InlinedItem,
                                           fold_ops: F)
                                           -> &'ast InlinedItem {
+    let _ignore = map.forest.dep_graph.in_ignore();
```
I would either add a comment here why this is needed, or just use `&map.forest.krate` instead of `map.krate()` below, so we don't register a read.

michaelwoerister commented on src/librustc/dep_graph/graph.rs:
```diff
@@ -70,6 +75,15 @@ impl DepGraph {
+    /// value is in scope. Note that this is only available when debug
+    /// assertions are enabled -- you should not check in code that
+    /// invokes this function.
+    #[cfg(debug_assertions)]
```
Nice. Ideally, I'd like to have an extended version of this in the future that's not just for ad hoc debugging.

michaelwoerister commented:
Good catches both!
r=me with the nits and addressed.

arielb1 commented:
> That said, the pattern that is used for the inherent impls map
is *generally* an anti-pattern (that is, holding the borrow lock for the
duration of using the contents)

I remember a similar place in coherence where I *intentionally* held the lock because someone (i.e., metadata loading) was inserting stuff to the list behind my back and I wanted to prevent that problem from recurring.

bors commented:
:umbrella: The latest upstream changes (presumably #36030) made this pull request unmergeable. Please resolve the merge conflicts.

nikomatsakis commented:
@bors r=mw

bors commented:
:pushpin: Commit 30c6f12 has been approved by `mw`

<!-- @bors r=mw 30c6f12877095043b717a1df2e17f5b2ca4d8bea -->

nikomatsakis commented:
@bors r-

nikomatsakis commented:
Warnings from travis look legit.

nikomatsakis commented:
@bors r=mw

bors commented:
:pushpin: Commit d287b7a has been approved by `mw`

<!-- @bors r=mw d287b7af1fa1017a9244389904c4f9122af9b87c -->

nikomatsakis commented:
@bors r- // again the travis issue looks legit, and again I wonder why I am not seeing this locally!

nikomatsakis commented:
@nagisa @michaelwoerister 

> Should this field also be #[cfg(debug_assertions)]-ed?

The answer is sort of yes, but I wound up (see most recent commit) taking a different approach. In particular, the number of things that must have `#[cfg]` applied to them seemed to be growing at a kind of ridiculous rate, and moreover making a field `#[cfg]` is kind of annoying since you can't apply `#[cfg]` to the field in the struct literals. I really find code with #[cfg] sprinkled all about in random places quite hard to read -- I like it to be contained at one or two "pinch points" or not used at all, so in the most recent commit I removed the `#[cfg(debug_assertions)]` on the `forbid` method and replaced it with an `assert!`. But I'm not totally happy with that.

I'm not sure what's a better way to formulate the code here. It occurs to me now I could maybe make a `mod forbid` that is different if debug-assertions are disabled, so that at least all the cond code can be contained in one place, but it still feels like a lot for such a small thing.

nikomatsakis commented:
>  again the travis issue looks legit, and again I wonder why I am not seeing this locally!

Oh, I am dumb, it's because I'm building with debug-assertions enabled, of course!

nikomatsakis commented:
OK, so, I wound up implementing a better mechanism altogether. I'll push a few more commits shortly.

nikomatsakis commented:
r? @michaelwoerister -- this fixes one add'l bug, and revamps the forbidden stuff into something more powerful.

bors commented:
:umbrella: The latest upstream changes (presumably #36203) made this pull request unmergeable. Please resolve the merge conflicts.

michaelwoerister commented on src/librustc/dep_graph/README.md:
```diff
@@ -392,6 +394,35 @@ This will dump out all the nodes that lead from `Hir(foo)` to
+    
+That first edge looks suspicious to you. So you set
+`RUST_FORBID_DEP_GRAPH_EDGE` to `Hir&foo -> Collect&bar`, re-run, and
+then observe the backtrace. Voila, bug fixed!
```
Nice! `:)`

michaelwoerister commented on src/librustc/dep_graph/shadow.rs:
```diff
@@ -0,0 +1,123 @@
+                    // we could easily check that the stack is
+                    // well-formed here, but since we use closures and
+                    // RAII accessors, this bug basically never
+                    // happens, so it seems not worth the overhead
```
The check that would be needed would just be that the top of the stack is the same DepNode as the field of PopTask, right? That doesn't sound like much of an overhead. I'd prefer just doing the check.

nikomatsakis replied:
> The check that would be needed would just be that the top of the stack is the same DepNode as the field of PopTask, right? That doesn't sound like much of an overhead. I'd prefer just doing the check.

OK. I originally had it doing the check but removed it for no particular reason. Just so that it had same overhead as today I guess.

michaelwoerister commented on src/librustc/dep_graph/shadow.rs:
```diff
@@ -0,0 +1,123 @@
+//! debug-assertions are not enabled.
+//!
+//! The basic sanity check, always enabled, is that there is always a
+//! task (or ignore) on the stack when you do read/write.
```
I don't see that in the diff (maybe in a later commit?).

nikomatsakis replied:
> I don't see that in the diff (maybe in a later commit?).

It is checked -- `check_edge` gets unhappy if the top of the stack is `None`, iirc.

michaelwoerister replied:
But it looks like control flow can never reach `check_edge` if debug assertions are turned off?

nikomatsakis replied:
> But it looks like control flow can never reach check_edge if debug assertions are turned off?

By "always enabled", I meant "always enabled if you build w/ debug assertions" (you don't need to e.g. use an env variable)

michaelwoerister replied:
OK, nevermind then `:)`

michaelwoerister commented on src/librustc/ty/mod.rs:
```diff
@@ -2437,12 +2437,41 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
+    ///
+    /// Note that if `id` is not local to this crate -- or is
+    /// inlined into this crate -- the result will be a non-local
+    /// `DefPath`.
```
Great documentation!

michaelwoerister commented:
@bors r+

bors commented:
:pushpin: Commit 9ca5786 has been approved by `michaelwoerister`

<!-- @bors r=michaelwoerister 9ca578687b88d2c7817d5709b2700fb6777348f2 -->

bors commented:
:hourglass: Testing commit 9ca5786 with merge fa9d8cc...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](https://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/10494), [auto-linux-32-opt](https://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/10548), [auto-linux-64-cargotest](https://buildbot.rust-lang.org/builders/auto-linux-64-cargotest/builds/1782), [auto-linux-64-cross-freebsd](https://buildbot.rust-lang.org/builders/auto-linux-64-cross-freebsd/builds/1607), [auto-linux-64-debug-opt](https://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/3766), [auto-linux-64-nopt-t](https://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/10489), [auto-linux-64-opt](https://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/10449), [auto-linux-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-linux-64-opt-rustbuild/builds/2421), [auto-linux-64-x-android-t](https://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/10426), [auto-linux-cross-opt](https://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/3647), [auto-linux-musl-64-opt](https://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/5598), [auto-mac-32-opt](https://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/10527), [auto-mac-64-opt](https://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/10501), [auto-mac-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-mac-64-opt-rustbuild/builds/2445), [auto-mac-cross-ios-opt](https://buildbot.rust-lang.org/builders/auto-mac-cross-ios-opt/builds/1625), [auto-win-gnu-32-opt](https://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/5615), [auto-win-gnu-32-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt-rustbuild/builds/2476), [auto-win-gnu-64-opt](https://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/5590), [auto-win-msvc-32-opt](https://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/4783), [auto-win-msvc-64-cargotest](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-cargotest/builds/1801), [auto-win-msvc-64-opt](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/5596), [auto-win-msvc-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-rustbuild/builds/2471)
Approved by: michaelwoerister
Pushing fa9d8cc8ace2c1f22469b4a6c8e1e0352072c675 to master...
